File: languages/haskell/flipper/src/Flipper/Internal/UART0.hs
{-|
Module      : Flipper.Internal.UART0
Description : Internal UART0 Module
Copyright   : George Morgan, Travis Whitaker 2016
License     : All rights reserved.
Maintainer  : travis@flipper.io
Stability   : Provisional
Portability : Windows, POSIX

-}

module Flipper.Internal.UART0 (
    configure
  , enable
  , disable
  , ready
  , put
  , get
  , push
  , pull
  ) where

import Data.Word

import Flipper.Internal.Buffer
import Flipper.Internal.Utils

import Foreign.ForeignPtr
import Foreign.Marshal.Utils
import Foreign.Ptr

configure :: IO Bool
configure = retSuc <$> c_uart0_configure

enable :: IO ()
enable = c_uart0_enable

disable :: IO ()
disable = c_uart0_disable

ready :: IO Bool
ready = toBool <$> c_uart0_ready

put :: Word8 -> IO ()
put = c_uart0_put

get :: IO Word8
get = c_uart0_get

push :: Buffer -> IO ()
push (Buffer p o l) = withForeignPtr p $ \p' ->
   c_uart0_push (plusPtr p' o) (fromIntegral l)

pull :: Int -> IO Buffer
pull l
    | l <= 0    = error "pull: length must be greater than zero."
    | otherwise = do b@(Buffer p _ _) <- allocBufferSafe l
                     withForeignPtr p (\p' -> c_uart0_pull p' (fromIntegral l))
                     return b

foreign import ccall safe "flipper/uart0/uart0.h uart0_configure"
    c_uart0_configure :: IO Word32

foreign import ccall safe "flipper/carbon/modules/uart0.h uart0_enable"
    c_uart0_enable :: IO ()

foreign import ccall safe "flipper/carbon/modules/uart0.h uart0_disable"
    c_uart0_disable :: IO ()

foreign import ccall safe "flipper/carbon/modules/uart0.h uart0_ready"
    c_uart0_ready :: IO Word8

foreign import ccall safe "flipper/carbon/modules/uart0.h uart0_put"
    c_uart0_put :: Word8 -> IO ()

foreign import ccall safe "flipper/carbon/modules/uart0.h uart0_get"
    c_uart0_get :: IO Word8

foreign import ccall safe "flipper/carbon/modules/uart0.h uart0_push"
    c_uart0_push :: Ptr Word8 -> Word32 -> IO ()

foreign import ccall safe "flipper/carbon/modules/uart0.h uart0_pull"
    c_uart0_pull :: Ptr Word8 -> Word32 -> IO ()

File: boards/carbon/hal/modules/uart0.c
#define __private_include__
#include <flipper/flipper.h>
#include <flipper/carbon/uart0.h>

int uart0_configure(void) {
	lf_invoke(&_uart0, _uart0_configure, NULL);
	return lf_success;
}

void uart0_enable(void) {
	lf_invoke(&_uart0, _uart0_enable, NULL);
}

void uart0_disable(void) {
	lf_invoke(&_uart0, _uart0_disable, NULL);
}

uint8_t uart0_ready(void) {
	return lf_invoke(&_uart0, _uart0_ready, NULL) >> 16;
}

void uart0_put(uint8_t byte) {
	lf_invoke(&_uart0, _uart0_put, fmr_args(fmr_infer(byte)));
}

uint8_t uart0_get(void) {
	return lf_invoke(&_uart0, _uart0_get, NULL) >> 16;
}

int uart0_push(void *source, lf_size_t length) {
	return lf_push(&_uart0, _uart0_push, source, length, NULL);
}

int uart0_pull(void *destination, lf_size_t length) {
	return lf_pull(&_uart0, _uart0_pull, destination, length, NULL);
}
